### ISSUE TITLE
[QST] why does Spark hang in local mode when I add GPU resources?

### ISSUE THREAD

**Issue by rohitreddy1698** [NVIDIA/spark-rapids#8651]
**What is your question?**
Hi , I am using a node which has only 1 GPU to try out spark RAPIDS in the local mode.

However the spark job is getting stuck whenever I try to read a parquet file when I use the following configuration : 
--conf spark.executor.resource.gpu.amount 1
--conf spark.rapids.sql.concurrentGpuTasks 2
--conf spark.task.resource.gpu.amount 0.5

But is working fine I use the following configuration :
--conf spark.executor.resource.gpu.amount 1

I understood from this thread : 
https://stackoverflow.com/questions/73277579/spark-rapids-does-not-load-unsupported-file-format-error-for-csv-and-no-error-f

that it is trying to assign 1 GPU per executor and 0.5 GPU to each task ; but my understanding is that when I say 
--conf spark.executor.resource.gpu.amount 1
--conf spark.rapids.sql.concurrentGpuTasks 2
--conf spark.task.resource.gpu.amount 0.5

It is assigning 1 GPU per executor out of which 0.5 will be used for 1 task and other 0.5 will be used for 1 task.

Is my Understanding wrong or can anyone help me with this please ?

**Comment by revans2** [comment]
It sounds like you are trying to run in local mode.  Spark has bugs in local mode when ever you try to assign resources and it hangs like you are seeing.

The one thing that you might want to be careful of is that if you set `spark.task.resource.gpu.amount` to `0.5` Spark will only assign 2 tasks per executor no matter how many cores are available. If that is all you want that is fine, but we have found that by having more CPU threads than tasks allowed on the GPU, we can get a lot better overlap of I/O and computation.

The I/O happens mostly on the CPU. `spark.rapids.sql.concurrentGpuTasks` only limits the number of tasks using the GPU at once.

**Comment by revans2** [comment]
Sorry I just saw that you said that it is indeed running in local mode.  Spark has bugs in local mode like I said.  Just try it without for now and then you can add them in when you run in a distributed setup.

**Comment by rohitreddy1698** [comment]
Hi , Apologies for the delay in my response.  I think I did not convey my question properly , let me rephrase it this way :

I have only 1 GPU on my system for starters , for the very initial testing phase : 

**Question 1)** In this configuration how much GPU would be used per executor ? - with this conf spark is unable to allocate resources and unable to read the parquet file
--conf spark.executor.resource.gpu.amount 1
--conf spark.rapids.sql.concurrentGpuTasks 2
--conf spark.task.resource.gpu.amount 0.5

**Question 2)** And how much GPU would be used in this configuration per executor ? - with this conf spark is able to read the parquet file
--conf spark.executor.resource.gpu.amount 1

From the stack over flow link I have pasted , this solution works for me :
"""
Most of the time a hang means that Spark could not allocate all of the resources needed to fulfill the resource request. Here you are running in local mode local[*] which means that Spark is going to try and allocate a task per CPU thread on your computer. But you have launched Spark with

>         --conf spark.executor.resource.gpu.amount=1 \
>         --conf spark.task.resource.gpu.amount=1 \
Which asks Spark to have 1 GPU per executor and 1 GPU per task. I am assuming that you have more cores than just one on your machine so Spark now is stuck it want to allocate X tasks, which will need X GPUs to run, but there is only 1 GPU available. Spark could be a lot better about throwing errors in these deadlock/miss-configuration cases.

When running in local mode you can only use 1 GPU. So the simplest way to launch it is to just remove all of the resource requests.
""""

**Question 3)** But my understanding was if I request 1 gpu per executor , that is to be shared between all the tasks running concurrently.

**Question 4)** Where the number of tasks per executor would be limited by : `spark.task.resource.gpu.amount` , irrespective of the value i set for : `spark.executor.cores` - from your explanation.

**Question 5)** And you are saying if I set `spark.executor.cores` more than `spark.task.resource.gpu.amount` - you have seen better performance.

But as per the explanation given in the stackoverflow thread , `spark.executor.resource.gpu.amount=1 ` - this is requesting 1 GPU resource and `spark.task.resource.gpu.amount` - this is requesting 1 GPU resource.



Regards,
Rohit Mothe

**Comment by tgravescs** [comment]
Are you running in local mode?  If so don't specify any of those configs and yes you can use the spark-rapids plugin and all the tasks will attempt to share that GPU.  At that point Spark doesn't do any scheduling based off of GPUs, purely off of CPUs and we just happen to use the GPU for Spark Rapids plugin.

1) breaks because you are saying each task needs part of a GPU and Spark doesn't support scheduling Gpus in local mode.  In other spark deployment modes where gpu scheduling is supported this would run a total of 2 tasks per executor.
2) works (I'm surprised this works) because you are saying your tasks don't actually need a GPU, so the same as not specifying. in local mode this might run as many cores as you specified.
3) correct, the GPU on an executor is shared among the tasks on that executor. normally you set spark.task.resource.gpu.amount to be 1/{number of cores on executor} to allow for it to run that the same number of tasks as cores.  Now, I don't recommend making the tasks per executor to large since you only have one GPU on that executor and the spark.rapids.sql.concurrentGpuTasks controls how many can be on the GPU at once.  Please see our documentation here: https://nvidia.github.io/spark-rapids/docs/FAQ.html#how-are-sparkexecutorcores-sparktaskresourcegpuamount-and-sparkrapidssqlconcurrentgputasks-related
4) sorry not sure what you are asking here?  Yes spark.task.resource.gpu.amount could limit the number of tasks per executor. Spark uses whatever is most restrictive.  Spark has documentation on resource aware scheduling at: https://spark.apache.org/docs/latest/configuration.html#custom-resource-scheduling-and-configuration-overview.

For example if I have executor.cores=8, spark.task.cpus=1, by cores I could have 8 tasks.  Now if I have spark.executor.resource.gpu.amount=1 and spark.task.resource.gpu.amount=1, this means that I can only run 1 task on that executor because GPU is the limiting factor.  Now if I change it and make spark.task.resource.gpu.amount=0.1, this means by GPU I could run 10 tasks, but then the cpus become the limiting factor and the executor would get 8 tasks.

5) I hope the example above answers this question.  I don't recommend setting spark.executor.cores to high, generally we keep it less then 16 but it depends some on hardware and use case.  This is some just normally Spark tuning, like you need to have fast enough disk drives for instance.

I think much of this confusion is coming from you running in local mode and Spark not really supposing gpu scheduling in local mode.  You could try running in standalone on a single server, where you start the master and worker on the same node and then Spark will support GPU scheduling properly.

**Comment by sameerz** [comment]
Closing this issue for now.  If you have any further questions feel free to re-open or start a https://github.com/NVIDIA/spark-rapids/discussions thread.
